bundle/config/mutator/translate_paths.go

```diff
@@ -124,91 +135,102 @@ func (m *translatePaths) translateFilePath(literal, localPath, remotePath string
-	var err error
+func (m *translatePaths) translateToBundleRootRelativePath(b *bundle.Bundle, literal, localPath, remotePath string) (string, error) {
+	return filepath.Rel(b.Config.Path, localPath)
+}
 
-	if task.NotebookTask != nil {
-		err = m.rewritePath(dir, b, &task.NotebookTask.NotebookPath, m.translateNotebookPath)
-		if target := (&ErrIsNotNotebook{}); errors.As(err, target) {
-			return fmt.Errorf(`expected a notebook for "tasks.notebook_task.notebook_path" but got a file: %w`, target)
-		}
-		if err != nil {
-			return err
-		}
-	}
+type transformer struct {
+	// A directory path relative to which `path` will be transformed
+	dir string
 
-	if task.SparkPythonTask != nil {
-		err = m.rewritePath(dir, b, &task.SparkPythonTask.PythonFile, m.translateFilePath)
-		if target := (&ErrIsNotebook{}); errors.As(err, target) {
-			return fmt.Errorf(`expected a file for "tasks.spark_python_task.python_file" but got a notebook: %w`, target)
-		}
-		if err != nil {
-			return err
-		}
-	}
+	// A path to transform
+	path *string
 
-	return nil
+	// Name of the config property where the path string is coming from
+	configPath string
+
+	// A function that performs the actual rewriting logic.
+	fn rewriteFunc
 }
+type transformerFactory func(*translatePaths, *bundle.Bundle) ([]*transformer, error)
+type selector struct {
+	// A path to transform
+	path *string
 
-func (m *translatePaths) translatePipelineLibrary(dir string, b *bundle.Bundle, library *pipelines.PipelineLibrary) error {
-	var err error
+	// Name of the config property where the path string is coming from
+	configPath string
 
-	if library.Notebook != nil {
-		err = m.rewritePath(dir, b, &library.Notebook.Path, m.translateNotebookPath)
-		if target := (&ErrIsNotNotebook{}); errors.As(err, target) {
-			return fmt.Errorf(`expected a notebook for "libraries.notebook.path" but got a file: %w`, target)
-		}
-		if err != nil {
-			return err
+	// A function that performs the actual rewriting logic.
+	fn rewriteFunc
+}
+type selectorFunc func(resource interface{}, m *translatePaths) *selector
+
+// List of available transformer factories. All of them will be called to get the list of transformers to execute.
+var transformerFactories []transformerFactory = []transformerFactory{
+	getJobsTransformers,
+	getPipelineTransformers,
+	getArtifactsTransformers,
+}
+
+// List of selector functions which are used to identify if the resource passed
+// Should have it's path transformed.
+// If it needs to be transformed, selector returns a reference to string to be transformed
+// And a function to apply for this string
+var selectors []selectorFunc = []selectorFunc{
+	selectNotebookTask,
+	selectSparkTask,
+	selectLibraryNotebook,
+	selectLibraryFile,
+	selectArtifactPath,
+	selectWhlLibrary,
+	selectJarLibrary,
+}
+
+// Appends the first matched transfomer for the given resource if it matches any if selectors defined
```

Comment:
I have trouble parsing what this means. It's not clear to me how selectors and transformers are different if you're building a transformer from the same fields as a selector.